clean up polygon intersection glitches VERSION 2 (using only shapely functions) (issue #2560)

jewettaijfc opened the pull request:
## This PR addresses issue #2560 

This diff uses several different functions from the `shapely` to cleanup polygons with thin or tiny hair-like protrusions.

Such features are often generated when computing intersections of polygons *(using `(ClipOperation.intersections_plane()`)*:  At certain rotation angles, polygons which would overlap exactly fail to be aligned due to floating-point rounding errors that occur during rotation.  This was causing visual glitches mentioned in [that issue](https://github.com/flexcompute/tidy3d/issues/2560), and shown below:

## *Before this PR*
```
box  = td.Box(center = (0,-9.01,10),
              size = (10,0.01,10))
cylinder = td.Cylinder(radius = 9,
                       center = (0,-9.01,0),
                       length=0.01,
                       axis = 1)
ax = (box-cylinder).rotated(np.pi/2,axis = 0).plot(x=0)
ax.set_aspect('auto') 
```
![Image](https://github.com/user-attachments/assets/c5d8d5ca-b045-4456-b174-6c9e439f752d)


## *After this PR*
![tidy3d_pr2581_after](https://github.com/user-attachments/assets/1d08b9a2-3062-4e64-8c00-57139697ee41)

NOTE: An earlier PR attempted to solve this problem [PR2581](https://github.com/flexcompute/tidy3d/pull/2581), but I decided to pursue this approach instead.


## Test files
Here is a file with more tests: 
[test_pr2596.ipynb.zip](https://github.com/user-attachments/files/20929827/test_pr2596.ipynb.zip)

(You must unzip it first.)

<!-- greptile_comment -->

## Greptile Summary

Implements polygon cleanup using `shapely.set_precision()` to eliminate thin triangles from polygon boundaries in 2D plots, offering a simpler but slower alternative to PR#2581.

- Added `cleanup_shapely_polygon()` function in `tidy3d/components/geometry/base.py` using shapely's built-in precision control
- Modified `ClipOperation.to_polygon_list()` with optional cleanup parameter for removing visualization artifacts
- Added comprehensive test suite in `tests/test_components/test_geometry.py` covering edge cases
- Performance impact.  The complicated benchmark example now requires 0.23 seconds, vs. 0.064s before this PR (or 0.1s using PR#2581).



<!-- /greptile_comment -->

jewettaijfc committed:
plots of objects defined by shape intersection logic no longer display thin line artifacts. VERSION 2 (issue #2560)

jewettaijfc commented:
I just found an example (described in [issue 2598](https://github.com/flexcompute/tidy3d/issues/2598)) causing the assertion I made in this PR to fail.  The first line of code can split a polygon into mulitple polygons:
```
cleaned_polygon = shapely.set_precision(shapely.Polygon(vertices).buffer(0), POLY_GRID_SIZE) 
assert isinstance(cleaned_polygon, shapely.Polygon)
```
...so the assertion on the next line line will fail in this case.  (In that example `cleaned_polygon` will be a `shapely.MultiPolygon`, not a `shapely.Polygon`.)  I need to add a few lines of code to deal with this issue.

jewettaijfc commented:
I fixed the previous issue.  It's okay to review this PR.

jewettaijfc commented:
### New failure type
I've been testing the PR for several hours today, and I found a more complicated example where both of approaches I've tried so far (PR2596 and PR2581) fail at certain rotation angles.
- Start with a 2x3 box.
- Remove a 1x2 box from it's lower-left corner ("bite1").
- Remove a 1x2 box from it's upper-right corner ("bite2").

![Image](https://github.com/user-attachments/assets/12ca16d3-d3f8-4688-8646-2146b7b0aa01)
```
box2x3_outer = td.Box(center = (1, 1.5, 0), size = (2,3,1))
box1x2_bite1 = td.Box(center = (0.5, 1, 0), size = (1,2,1))
box1x2_bite2 = td.Box(center = (1.5, 2, 0), size = (1,2,1))

split_box = (box2x3_outer - box1x2_bite1) - box1x2_bite2

split_box.rotated(0.23*np.pi/2, axis = 2).plot(z=0)
```
You *should* see two squares.  Unfortunately, this is what we sometimes see:

![tidyed_pr2596_failure1](https://github.com/user-attachments/assets/f1ee88a1-9dcd-4f01-8f51-b124adf2f602)
![tidy3d_pr2596_failure2](https://github.com/user-attachments/assets/65686492-2649-4146-9dfd-d2204853f980)

I tried increasing the grid size used for geometry simplification, but it did not help that much for this example.  There were always some angles where this example had visual glitches.

## Possible issues with our existing code?


As Weiliang points out, I am currently using [the same approach we use elsewhere in Tidy3D](https://github.com/flexcompute/tidy3d/pull/2581#issuecomment-2993325491) to cleanup polygons:
```
shapely.set_precision(shapely.Polygon(vertices).buffer(0), POLY_GRID_SIZE)
```
This trick helps most of the time, but fails in the example above.

**UPDATE: I no longer use this approach.  (See below.)**

jewettaijfc commented:
## UPDATE: I fixed this problem by changing the approach

To solve these problems, I now use buffering (a.k.a. "erosion", "dilation") with *non-zero distance* parameters (ie 1e-12).

What that means in simple language:

1) First I "erode" (srhink) the polygons by a small distance (1e-12) to get rid of outward spikes.
2) Then I "dilate" (expand) the (previously eroded) polygons (by a ***larger*** distance 2e-12) to get rid of inward spikes.
3) Then I erode/shrink the polygons again (by 1e-12) back to the original size.

Here is a simplified version of the relevant code.
```
eroded = shapely.buffer(original_polygon, distance=-resolution)  # removes outward spikes
dilated = shapely.buffer(eroded, distance=2 * resolution)  # removes inward spikes
cleaned = shapely.buffer(dilated, distance=resolution)  # expands it back to the original size
```

This will get rid of spikes that protrude either *outward or inward*, and it will handle changes in topology like the two-square example above.

This seems to work really well.

(Perhaps we should update the rest of the other polygon-cleaning code to do it this way?)

jewettaijfc commented:
I just noticed the unit test failures.  I'll try and fix that in the next hour.

*(UPDATE: "next hour"   ha ha)*

jewettaijfc commented:
I finally fixed all of the issues which were causing unit test failures.  (Lots of failures!)
This PR is ready for review again!

jewettaijfc commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3319,40 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    )
+    # Assertion fails when polygon intersects itself (eg: ((0,0), (1,1), (2,-1), (3,1), (4,0)))
+    assert isinstance(cleaned_polygon, shapely.Polygon)
+    return cleaned_polygon
```
## *UPDATE: THIS CODE EXCERPT IS OUTDATED.*
Please ignore that code excerpt.  (Sorry.  A lot of changes have been made since this PR was first posted.)

yaugenst-flex commented on tests/test_components/test_geometry.py:
```diff
@@ -211,6 +214,10 @@ def test_intersections_plane(component):
 
 
+def test_intersections_plane_DELETEME():  # DEBUG - REMOVE THIS FUNCTION BY 2025-6-25 -Andrew
+    assert len(SYM_DIFFERENCE.intersections_plane(z=0.2)) > 0
```
Delete I guess?

jewettaijfc replied:
Thanks for catching that!

yaugenst-flex commented on tests/test_components/test_geometry.py:
```diff
@@ -14,6 +14,9 @@
 
 import tidy3d as td
+from tidy3d.components.geometry.base import (
+    cleanup_shapely_object,
```
nit: remove comma after `cleanup_shapely_object`

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3319,92 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    )
+    # Important: Remove any self intersections beforehand using `shapely.make_valid()``
+
+    try:
+        valid_obj = shapely.make_valid(normalized_obj, method="structure", keep_collapsed=False)
+    except (TypeError, AttributeError):
+        # (Old versions of shapely don't support `shapely.make_valid()`.  Use buffer(0) instead.)
+        valid_obj = normalized_obj.buffer(0)
```
This seems strange. As far as I can tell, `.make_valid()` was introduced in shapely `1.8a1 (2021-03-03)`, and we pin `shapely = "^2.0"`. Could you double check whether this is really the reason that the tests fail on python 3.9? You can `poetry env use python3.9` to set up your env with a specific python version.

yaugenst-flex replied:
Same comment for the tests too.

jewettaijfc replied:
Thanks for telling me about `poetry env use python3.9`.  I tried it and I'm fighting with conda and poetry.lock file conflicts right now.  But it's a good learning experience for me.  Hold on while I figure this out...

jewettaijfc replied:
Okay.  It looks like this is necessary.
- Although `make_valid()` has been around since `1.8a1 (2021-03-03)`, the `method` and 'keep_collapsed' arguments were not added to the `make_valid()` function until shapely version `2.1.0`.
- In a python 3.9 environment, when I use `poetry install -E dev` inside the `tidy3d` directory, it installs `shapely.__version__='2.0.7'`.  *(That raises a `TypeError`.  Even older versions of shapely/python that don't support `make_valid()` will throw an `AttributeError`.)*

**Background information:**

`buffer(0)` is deprecated and `make_valid()` is it's replacement.   The `method="structure"` and `keep_collapsed=False`arguments simplify the output of the `make_valid()` function, but I suppose we don't have to use these arguments.  Either way, I attempted to harden the code so that nothing terrible will happen, regardless whether we use `make_valid()` or `buffer(0)`.

yaugenst-flex replied:
I see. If the arguments are not necessary then I think the cleaner approach would be to just not use them, or would that significantly complicate the code? Or I guess, maybe the other way round is worth considering too: Is `make_valid()` necessary, i.e., does it add significant value beyond just `buffer(0)`? Because if it doesn't, we could just use that instead.

jewettaijfc replied:
***If*** we decide to use `make_valid()`, then then we should use those arguments *(`method="structure"` and `keep_collapsed=False`)* to maintain type consistency.  If my understanding is correct, these arguments ensure that `make_valid()` will only return objects of type `shapely.Polygon` or `shapely.MultiPolygon` (as does `buffer(0)`).  Otherwise [make_valid() might return `shapely.Point` or `shapely.LineString`, or other 1-D or 0-D objects](https://shapely.readthedocs.io/en/stable/reference/shapely.make_valid.html).  (After my recent code changes, *I think* that would still be okay.  But I have seen tidy3d break in unexpected places when that happens.)

If we want a cleaner, simpler solution, then we can just use `buffer(0)` inside  `cleanup_shapely_object()` function as a replacement for `make_valid()`.  Regarding that issue, this is what gemini.google.com told me:

*"It's important to note that while buffer(0) is a common "trick" for cleaning, it's not always foolproof for all types of invalid geometries. For more robust validation and repair, shapely.make_valid() is often preferred in newer shapely versions (Shapely 2.0+)."*

I have noticed that `buffer(0)` is more likely to generate polygons with duplicate vertices.  But these are visually harmless.   However `buffer(0)` *does not* work in `test_cleanup_shapely_object()`, because that unit test is very sensitive to the number of vertices in the final polygon.  So I skip that test when `make_valid()` is not available.  (Thanks for showing me how to do that, btw.)

jewettaijfc replied:
I decided to go with `buffer(0)`.

yaugenst-flex replied:
ok i see. i guess ideally we'd want to pin `shapely>=2.1` then but it seems they dropped support for python 3.9, which we'll keep supporting until october..
i think it's fine as-is then for now. probably not great to default to a less robust method. the only thing i'd consider is wrapping `make_valid` for `shapely<2.1` in `tidy3d/compat.py` and importing from there. makes it easier to remember to remove later once we drop support for python 3.9, and we won't have to touch `cleanup_shapely_object`. so something like:
```py
# compat.py

from importlib.metadata import version

from packaging.version import parse as parse_version
from shapely import make_valid as _make_valid_base

if parse_version(version("shapely")) >= parse_version("2.1"):
    make_valid = _make_valid_base
else:
    def make_valid(geom, *arg, **kwargs):
        return geom.buffer(0)
```

```
# base.py

from tidy3d.compat import make_valid
```

jewettaijfc replied:
Thanks Yannick!  I'll do that now.  Hold on..

jewettaijfc replied:
> Thanks Yannick! I'll do that now. Hold on.

This is a really cool trick by the way.
But the code breaking in weird places if I do that.
But I kind of want to pursue why it's doing that.  Because it shouldn't be.

jewettaijfc replied:
After I implemented this change to the compat.py file, we now have a failing unit test.(`test_validate_polyslab_vertices_valid()`.  Recall that this change replaces `shapely.make_valid()`  with `buffer(0)` when python 3.9 is in use).

I don't know why it failed this time *(and not previously when i was also using buffer(0))*.

***But I'm glad it did.***

It convinced me that perhaps `buffer(0)` is not very robust.  Perhaps I should stop using it inside `cleanup_shapely_object()`.

Getting `cleanup_shapely_object()` to work with python3.9 has taken me an entire day of work.  (It's not your fault.  Both `make_valid()` and `buffer(0)` are really finicky.)  Do you want me to keep working on this?

I'm inclined to disable `cleanup_shapely_object()` when using shapely version < 2.1 (python3.9).  If our users complain about visual glitches in their plots, they can upgrade to python 3.11.

jewettaijfc replied:
>  I'm inclined to disable cleanup_shapely_object() when using shapely version < 2.1 (python3.9).

Done.
(Thanks for letting me know we are no longer supporting python 3.9, btw.  That gave me the courage to make this decision.)

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3319,92 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
         return grad_vjps
 
 
+def cleanup_shapely_object(obj: Shapely, resolution: float = POLY_GRID_SIZE) -> Shapely:
```
The function assumes that `obj` is not empty, not sure if the function would crash if it is? Might be good to add a guard like
```py
if obj.is_empty:
  return obj
```

yaugenst-flex commented on tests/test_components/test_geometry.py:
```diff
@@ -1249,3 +1256,59 @@ def wrong_shape_height_func(x, y):
+    exterior_coords = triangle_collinear  # has zero area
+    orig_polygon = shapely.Polygon(exterior_coords)
+    new_polygon = cleanup_shapely_object(orig_polygon, resolution=1e-12)
+    try:
+        new_polygon = shapely.make_valid(
+            shapely.set_precision(new_polygon, 1e-10),
+            method="structure",
+            keep_collapsed=False,
+        )
+    except (TypeError, AttributeError):
+        # (Old versions of shapely don't support `shapely.make_valid()`.)
+        return
```
This will never be reached if `shapely.make_valid()` errors. The comment below regarding `shapely.make_valid()` notwithstanding, this should probably be revisited.

yaugenst-flex commented on tests/test_components/test_geometry.py:
```diff
@@ -1249,3 +1256,59 @@ def wrong_shape_height_func(x, y):
+            method="structure",
+            keep_collapsed=False,
+        )
+    except (TypeError, AttributeError):
+        # (Old versions of shapely don't support `shapely.make_valid()`.)
+        return
```
Don't `return` here but instead [`pytest.skip`](https://docs.pytest.org/en/stable/how-to/skipping.html#skipping-test-functions) with the reason, otherwise it looks like the test passed.

jewettaijfc replied:
Awesome.  Thanks for the suggestion.  I'll keep that in mind for the future.

I just realized that I can put
```
new_polygon = new_polygon.buffer(0)
```
here, and that also works fine.  *(I was afraid the test would fail if I did that, but it passes.)*

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -2862,27 +2862,47 @@ def _geometries_untraced(cls, val):
 
     @staticmethod
-    def to_polygon_list(base_geometry: Shapely) -> list[Shapely]:
+    def to_polygon_list(base_geometry: Shapely, cleanup_afterwards: bool = False) -> list[Shapely]:
```
I think the `afterwards` here is an implementation detail that the caller of the function does not need to know about. I'd suggest just naming this `cleanup`.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -2862,27 +2862,47 @@ def _geometries_untraced(cls, val):
         List[shapely.geometry.base.BaseGeometry]
             Valid polygons retrieved from ``base geometry``.
         """
+        unfiltered_geoms: list[Shapely] = []
```
We typically don't type hint within functions. Doesn't hurt I guess.

jewettaijfc replied:
I'll try to avoid them in the future.  But in this particular example, I like the type hint.  *(It clarifies that `unfiltered_geoms` is a list of general `Shapely` objects.  I remember I was a little bit confused about that detail when I first read the code for this function.  At first, I thought everything was of type `shapely.Polygon`.)*

But if you want me to remove these type hints here too, I'm happy to do that.

jewettaijfc replied:
removed.

yaugenst-flex commented on tidy3d/components/viz/descartes.py:
```diff
@@ -35,7 +35,10 @@ def __init__(self, context):
     def exterior(self):
         """Get polygon exterior."""
-        return getattr(self.context, "exterior", None) or self.context[0]
+        value = getattr(self.context, "exterior", None)
+        if value is None:
+            value = self.context[0]
+        return value
```
Why change this?

jewettaijfc replied:
The `exterior` and `coding` functions in the `decartes.py` file were crashing whenever the polygon (self.context.exterior) was empty.  *(See "Details" below.)*

I noticed this because the `cleanup_shapely_object()` function (introduced in this PR) can remove all of the vertices from a polygon, resulting in empty polygons..

This problem happens rarely.  (The unit tests did not catch it.)  But at some point it was crashing here when I ran one of the complex examples from `tidy3d-notebooks`.  That's why I updated both `exterior` and `coding` functions so they can handle empty polygons better.

-----------------------------------------
**Details:**

This happens when `self.context` is an empty `shapely.Polygon`.  Let's take a look at the original function:
```
    @property
    def exterior(self):
        return getattr(self.context, "exterior", None) or self.context[0]
```

If `self.context.exterior` is empty, it will evaluate to `False` when inside an `or` statement, even though `self.context.exterior` is not the same as `None`.  This returns `self.context[0]` which crashes since `self.context` is of type `shapely.Polygon`.  I don't think this was the intended behavior.

Incidentally, I am imitating [the code from the existing `interiors` function](https://github.com/flexcompute/tidy3d/blob/994bdc253041a9fd1a887b315ed16a38012cf3a5/tidy3d/components/viz/descartes.py#L43) here.

yaugenst-flex replied:
ah ok, yeah that makes sense! thanks

yaugenst-flex commented on tidy3d/components/viz/descartes.py:
```diff
@@ -53,9 +56,13 @@ def polygon_path(polygon):
         # The codes will be all "LINETO" commands, except for "MOVETO"s at the
         # beginning of each subpath
-        n = len(getattr(obj, "coords", None) or obj)
+        crds = getattr(obj, "coords", None)
+        if crds is None:
+            crds = obj
+        n = len(crds)
```
Same here. Looks like the previous implementation was fine?

jewettaijfc replied:
It was not.  This was brittle.  See my comment above.

yaugenst-flex commented on tidy3d/components/viz/descartes.py:
```diff
@@ -53,9 +56,13 @@ def polygon_path(polygon):
+        n = len(crds)
         vals = ones(n, dtype=Path.code_type) * Path.LINETO
-        vals[0] = Path.MOVETO
+        if len(vals) > 0:
+            vals[0] = Path.MOVETO
```
Can you explain this change?

jewettaijfc replied:
Sure!  It's related to the other changes I made in this file.
- The new `cleanup_shapely_object()` function can remove all of the vertices from a polygon, increasing the risk that `vals` could be empty.
- So I inserted a `if len(vals) > 0:` check before the `vals[0] = Path.MOVETO`.

yaugenst-flex reviewed (commented):
Thanks @jewettaijfc this look pretty good overall! Left a couple of comments but most of the points are relatively minor. I think the performance impact is justifiable, since it only concerns plotting. Is there any worst-case scenario where the slowdown is really noticeable?

jewettaijfc commented:
Thanks for all the comments!

The things you thought were weird are weird.  Those changes were added yesterday to get around all the strange little problems I encountered.

I am addressing your comments now.  I'll comment as I go through each one..

jewettaijfc commented:
This PR is ready for review again.

(Ignore the unit test failures.  Our testing infrastructure has been having issues with docker all day today.)

weiliangjin2021 commented:
> Polygon + rotation really poses lots of challenges!

Yeah!

>> I tried increasing the grid size used for geometry simplification, but it did not help that much for this example.
> From performance perspective, is `set_precision` faster? In my test, `set_precision` works with grid size 1e-10. If it's just for plotting, a grid size around single float epsilon 1e-7 should be sufficient enough.

I haven't tested the speed of `set_precision()`.  I suspect it is very fast.  But I no longer use `set_precision()` in the main code for this PR.

### Speed comparison
Fortunately, the extra code I added since your suggesttion did not slow things down that much.

The last cell of [**this jupyter file**](https://github.com/user-attachments/files/20929827/test_pr2596.ipynb.zip) contains a benchmark example.  (You must unzip it first.)
On my laptop, after my most recent changes
- [**THIS PR 2596**](https://github.com/flexcompute/tidy3d/pull/2596) now requires 0.23 seconds.
- Previously it required 0.195 seconds. *(It was using `set_precision(polygon.buffer(0), POLY_GRID_SIZE)`.  I suspect most of the computation time was due to `buffer(0)`)*
- The `develop` branch requires 0.064 seconds to draw the plot in the benchmark example
- [**PR 2581**](https://github.com/flexcompute/tidy3d/pull/2581) requires 0.098 seconds.  *(But it was not a robust solution, so I closed that PR.  Indeed, `shapely` functions are more robust.)*

*(All times reported are the best/fastest time after 10 attempts.)*

**Why I don't use set_precision()**

`set_precision()` is very helpful to get rid of overlapping vertices.  It also helps for rectangular polygons which are aligned with the x,y,z axes.  But it's not as useful for vertices on lines at arbitrary angles.  The [second example](https://github.com/flexcompute/tidy3d/pull/2596#issuecomment-2998424331) (after rotation) has vertices along the line that are *almost* collinear, after rotation.  Using `set_precision()` did not help in that case because the line was not pointing in the x, y, or z direction.

I still use `set_precision()` in the `text_cleanup_shapely_object()` function, because the square polygon from that test example is aligned with the x,y,z axes.

Thank you again, Weiliang for being part of this discussion!  In addition to `set_precision()`, you also suggested I look at `buffer()`.  This is what saved me.  I use `buffer()` in this PR.  Your comments encouraged me to take a serious look at what shapely provides.

jewettaijfc commented:
## Removed support for pretty plots in python 3.9
(More specifically, I removed support for versions of `shapely` below 2.1.  Shapely 2.1 is incompatible with python 3.9.)

Getting this PR working with python 3.9 was taking too long.
(Thanks again, Yannick for telling me we will no longer be supporting 3.9 soon.)

jewettaijfc commented:
## This PR is ready for review again.

jewettaijfc commented:
> ## Removed support for python 3.9
> (More specifically, I removed support for versions of `shapely` below 2.1. Shapely 2.1 is incompatible with python 3.9.)
> 
> Getting this PR working with python 3.9 was taking too long. (Thanks again, Yannick for telling me we will no longer be supporting 3.9 soon.)

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+            " to v2.1) may cause plot errors.  Consider upgrading to\n"
+            " python 3.11 (or later) and reinstalling tidy3d.",
+            log_once=True,
+        )
```
The message is not quite accurate, it only depends on the shapely version no?

yaugenst-flex replied:
The message is not quite accurate, it only depends on the shapely version no?

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+            " python 3.11 (or later) and reinstalling tidy3d.",
+            log_once=True,
+        )
+        return obj  # Do not attempt to modify the object.
```
Remove comment

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -63,6 +66,7 @@
 from .bound_ops import bounds_intersection, bounds_union
 
 POLY_GRID_SIZE = 1e-12
+SHAPELY_IS_OLD = parse_version(importlib.metadata.version("shapely")) < parse_version("2.1")
```
I'd prefer moving this logic into `tidy3d/compat.py`

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -2862,27 +2866,47 @@ def _geometries_untraced(cls, val):
+            If True, then extremely thin or tiny features on the polygon's boundary will be removed.
+            (Vertices will be rounded to the nearest grid point with spacing ``POLY_GRID_SIZE``,
+            and any zero-area triangles that result will be deleted.)  This is useful for removing
+            artifacts from 2D plots displayed to the user.
```
Let's make this more declarative:
```
If True, removes extremely small features from the polygon's boundary. It achieves this by rounding vertices to the nearest grid point, defined by the ``POLY_GRID_SIZE`` spacing, and subsequently deleting any resulting zero-area triangles. This process is effective for eliminating visual artifacts in 2D plots.
```

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    Parameters
+    ----------
+    obj : shapely
+        a shapely object (typically a Polygon or a MultiPolygon)
```
` ``Polygon`` ` and ` ``MultiPolygon`` `

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+        a shapely object (typically a Polygon or a MultiPolygon)
+    resolution : float = ``POLY_GRID_SIZE``
+        Features that are smaller than ``resolution`` (relative to the object's size)
+        will be discarded.
```
Reading this it's not immediately obvious to me what `(relative to the object's size)` means here.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    Notes
+    -----
+    This function does not attempt to delete overlapping, nearby, or collinear vertices.
+    To solve that problem, use shapely.simplify() afterwards.  (Warning: This can be slow.)
```
` ``shapely.simplify()`` ` and remove `(Warning: This can be slow.)`

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+        return obj
+    centroid = obj.centroid
+    object_size = max(obj.bounds[2] - obj.bounds[0], obj.bounds[3] - obj.bounds[1])
+    if object_size == 0.0:  # edge case
```
Unnecessary comment, remove.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+        return shapely.Polygon([])
+    # In order to prevent numerical overflow or underflow errors, we first subtract
+    # the centroid and divide by (rescale) the size of the object so it is not too big.
+    # (If we don't do this then some of the unit tests with huge polygons will fail.)
```
Remove `()` comment

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+            0.0,
+            0.0,
+            1 / object_size,  # multiply y coordinates by 1 / object_size
+            -centroid.x / object_size,  # subtract the centroid (after dividing by object_size)
```
These comments are duplicates of what was already explained in the comment block above. Let's keep one or the other.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    # Important: Remove any self intersections beforehand using `shapely.make_valid()`.
+    valid_obj = shapely.make_valid(normalized_obj, method="structure", keep_collapsed=False)
+    # To get rid of small thin features, erode(shrink), dilate(expand), and erode again.
+    # (Alternatively, you could dilate, erode, and dilate again.  That works too.)
```
Remove `()` comment, serves no purpose in the context of the code here.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+    # Optional: Now shrink the polygon back to the original size.
+    # Note: This is not really necessary, and we could reduce computation time by skipping it.
+    # However I decided to do it anyway.  I do it because am paranoid weird things could happen
+    # with other parts of the code which make assumptions about the polygon's size and topology.
```
Can we condense this comment into a single line? Most of it isn't adding much information for whoever is reading the code.

yaugenst-flex commented on tidy3d/components/geometry/base.py:
```diff
@@ -3293,4 +3323,99 @@ def _compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradField
+        quad_segs=3,
+    )
+    # Earlier we subtracted the centroid and divided by the size of the object.
+    # Now we undo that.
```
More concise: `# Revert to the original scale and position.`

yaugenst-flex reviewed (commented):
Thanks @jewettaijfc, this is getting very close! For the most part only the comments need some polish. Generally comments should be concise and add necessary information for someone unfamiliar with the code, and avoid adding noise (the code itself is "self-documenting" to some extent).

yaugenst-flex commented on tests/test_components/test_geometry.py:
```diff
@@ -1249,3 +1252,50 @@ def wrong_shape_height_func(x, y):
+
+
+def test_cleanup_shapely_object():
+    if parse_version(importlib.metadata.version("shapely")) < parse_version("2.1"):
```
This should probably reuse what we implemented for `compat`

yaugenst-flex commented on tidy3d/compat.py:
```diff
@@ -2,9 +2,32 @@
 
 from __future__ import annotations
 
+# xarray
```
remove

yaugenst-flex commented on tidy3d/compat.py:
```diff
@@ -2,9 +2,32 @@
 except ImportError:
     from xarray.core import alignment
 
+# shapely
```
remove

yaugenst-flex commented on tidy3d/compat.py:
```diff
@@ -2,9 +2,32 @@
     from xarray.core import alignment
 
+# shapely
+import importlib
+
+from packaging.version import parse as parse_version
+
+from tidy3d.log import log
```
imports to the top

yaugenst-flex commented on tidy3d/compat.py:
```diff
@@ -2,9 +2,32 @@
+    if _SHAPELY_VERSION < parse_version(version):
+        log.warning(
+            "Using old versions of the shapely library (prior to v2.1) may cause\n"
+            "plot errors.  This can be solved by upgrading to python 3.11\n"
```
replace `\n` with a space. Superfluous space in `errors.  This`

Python 3.11 should be 3.10

also, reword: "Using versions versions of shapely prior to 2.1 [...]"

`tidy3d -> Tidy3D`

`python -> Python`

also, i'd actually not mention reinstalling tidy3d. i don't even know how you would do the upgrade without having to reinstall everything

yaugenst-flex reviewed (commented):
Thanks, almost there!

jewettaijfc commented:
(I guess our testing infrastructure is overwhelmed during normal working hours.  So I ran the unit tests manually using `poetry run pytest --tb=long  tests` using python 3.9 just now.  All unit tests are passing for both 3.11 and 3.9.)
